File: python/Tree/Binary_Search_Tree_Operations.py
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


def search(root, key):
  if not root or root.val == key:
    return root
  
  if key < root.val:
    return search(root.left, key)

  return search(root.right, key)

def insert(root, key):
  if not root:
    return TreeNode(key)
  
  if key < root.val:
    root.left = insert(root.left, key)
  else:
    root.right = insert(root.right, key)
  return root

def delete(root, key):
  if not root:
    return None 
  
  if key < root.val:
    root.left = delete(root.left, key)
  
  elif key > root.val:
    root.right = delete(root.right, key)
  
  else:
    if not root.left:
      return root.right
    if not root.right:
      return root.left
    temp = root.right
    while temp.left:
      temp = temp.left
    root.val = temp.val
    root.right = delete(root.right, temp.val)
    
  return root

File: C++/Tree/Binary_Search_Tree_Operations.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// Insert
TreeNode *insertIntoBST(TreeNode *root, int val)
{
  if (!root)
    return new TreeNode(val);

  if (val < root->val)
    root->left = insertIntoBST(root->left, val);
  else
    root->right = insertIntoBST(root->right, val);
  return root;
}

// Search
TreeNode *searchBST(TreeNode *root, int val)
{
  if (!root || root->val == val)
    return root;

  if (val < root->val)
    return searchBST(root->left, val);

  return searchBST(root->right, val);
}

// Delete
TreeNode *findMin(TreeNode *root)
{
  while (root->left)
    root = root->left;

  return root;
}

TreeNode *deleteNode(TreeNode *root, int key)
{
  if (!root)
    return root;

  if (key < root->val)
    root->left = deleteNode(root->left, key);
  else if (key > root->val)
    root->right = deleteNode(root->right, key);
  else
  {
    if (!root->left)
      return root->right;
    if (!root->right)
      return root->left;

    TreeNode *temp = findMin(root->right);
    root->val = temp->val;
    root->right = deleteNode(root->right, temp->val);
  }
  return root;
}